core/src/main/java/io/grpc/internal/CompositeReadableBuffer.java

```diff
@@ -147,46 +161,79 @@ public void readBytes(OutputStream dest, int length) throws IOException {
     execute(STREAM_OP, length, dest, 0);
   }
 
+  private static final NoThrowReadOperation<CompositeReadableBuffer> COMPOSITE_OP =
+      new NoThrowReadOperation<CompositeReadableBuffer>() {
+        @Override
+        public int read(ReadableBuffer buffer, int length, CompositeReadableBuffer dest,
+            int unused) {
+          dest.addBuffer(buffer.readBytes(length));
+          return 0;
+        }
+      };
+
   @Override
   public ReadableBuffer readBytes(int length) {
-    if (length <= 0) {
-      return ReadableBuffers.empty();
+    final CompositeReadableBuffer newBuffer = new CompositeReadableBuffer();
```

Comment:
Unfortunately, this reverts the optimization made in #7375. We cannot directly take the buffer instance from the source composite, as it should not retain the old mark position after the read.